frontend/src/components/editor/Editor.tsx

```diff
@@ -25,11 +30,158 @@ function Editor() {
 	const workspaceStore = useSelector(selectWorkspace);
 	const { mutateAsync: createUploadUrl } = useCreateUploadUrlMutation();
 	const { mutateAsync: uploadFile } = useUploadFileMutation();
+
+	const [showFormatBar, setShowFormatBar] = useState(false);
+	const [formatBarPosition, setFormatBarPosition] = useState({ top: 0, left: 0 });
+	const [selectedFormats, setSelectedFormats] = useState<Set<FormatType>>(new Set());
+	const [debouncedUpdate, setDebouncedUpdate] = useState<null | ViewUpdate>(null);
+
 	const ref = useCallback((node: HTMLElement | null) => {
 		if (!node) return;
 		setElement(node);
 	}, []);
 
+	const getFormatMarker = useCallback((formatType: FormatType) => {
+		switch (formatType) {
+			case FormatType.BOLD:
+				return "**";
+			case FormatType.ITALIC:
+				return "_";
+			case FormatType.CODE:
+				return "`";
+			case FormatType.STRIKETHROUGH:
+				return "~~";
+		}
+	}, []);
+
+	const getFormatMarkerLength = useCallback((state: EditorState, from: number) => {
+		const maxCheckLength = 10;
+		const docSlice = state.sliceDoc(Math.max(0, from - maxCheckLength), from).toString();
+		let cnt = 0;
+
+		for (let i = docSlice.length - 1; i >= 0; i--) {
+			if (!["*", "_", "`", "~"].includes(docSlice[i])) break;
+			cnt++;
+		}
```

Comment:
How about change it to `reduce`?
I want to aviod using index operation for readability.

```typescript
const cnt = docSlice.reduce((acc, c) => {
    if (["*", "_", "`", "~"].includes(c)) ++acc;
    return acc;
}, 0);
```